nemo/utils/exp_manager.py

```diff
@@ -0,0 +1,213 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import subprocess
+import sys
+import time
+from dataclasses import dataclass
+from pathlib import Path
+from shutil import copyfile
+from typing import List, Optional, Union, Dict
+
+from omegaconf import OmegaConf, DictConfig
+from hydra.utils import get_original_cwd
+from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.loggers import TensorBoardLogger
+
+from nemo.constants import NEMO_ENV_VARNAME_DATETIME
+from nemo.utils import logging
+from nemo.utils.get_rank import is_global_rank_zero
+
+
+@dataclass
+class ExpManagerConfig:
+    """TODO: Is this how you document a dataclass?
+    Args:
+        root_dir (str, Path): The base directory to create the logging directory. Defaults to None, which logs to
+            ./NeMo_experiments.
+        name (str): The name of the experiment. Defaults to None, which uses the lightning default of "default".
+        create_tensorboard_logger (bool): Whether to create a tensorboard logger and attach it to the pytorch lightning
+            trainer. Defaults to True.
+        create_checkpoint_callback (bool): Whether to create a ModelCheckpoint callback and attach it to the pytorch
+            lightning trainer. The ModelCheckpoint saves the top 3 models with the best "val_loss" as well as the most
+            recent model. Defaults to True.
+        files_to_copy (list): A list of files to copy to the experiment logging directory. Defaults to None which copies
+            no files.
+    """
+
+    root_dir: Optional[str] = None
+    name: Optional[str] = None
+    create_tensorboard_logger: Optional[bool] = True
+    create_checkpoint_callback: Optional[bool] = True
+    files_to_copy: Optional[List[str]] = None
+
+
+def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictConfig, Dict]] = None):
+    """
+    exp_manager is a helper function used to manage folders for experiments. It follows the pytorch lightning paradigm
+    of root_dir/model_or_experiment_name/version. If the lightning trainer has a logger, exp_manager will get root_dir,
+    name, and version from the logger. Otherwise it will use the root_dir and name arguments to create the logging
+    directory. The version will be a datetime string if running single node, and version will be an integer if running
+    on slurm multi-node.
+    It optionally creates TensorBoardLogger, and ModelCheckpoint objects from pytorch lightning. It copies sys.argv,
+    and git information if available to the logging directory. It creates a log file for each process to log their
+    output into.
+
+    Args:
+        trainer (pytorch_lightning.Trainer): The lightning trainer.
+        cfg (ExpManagerConfig)
```

Comment:
Please copy the doctoring for the supported parameters in the dataclass down to this docstring. The user facing code should have the docstrimg, not the implementation detail dataclass v

Reply:
How should I format it?

Reply:
```
  ...
  cfg (DictConfig): Can have the following optional keys
    - root_dir (str) : ... 
    - name (str): ...
    - ...
```

And inside the code, use cfg.get(<name of arg>, <default value of arg>). Then the dataclass is entirely redundant c

Reply:
I will update the docstring,

Yes but the `cfg.get(<name of arg>, <default value of arg>)` has the same issue as you pointed out below, it does not catch typos. Which is why I believe dictionaries, and by extension phase 1, is a bad idea and not should be used for passing arguments.

Reply:
Sounds good. And since the merge catches these issues, this is a good use-case (when its an internal one off check at least).